clash-cores/test/Test/Cores/Xilinx/DcFifo.hs

```diff
@@ -0,0 +1,139 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Test.Cores.Xilinx.DcFifo ( tests ) where
+
+import qualified Prelude as P
+import qualified Data.List as L
+import Data.Maybe (catMaybes, fromMaybe)
+import Control.Monad (replicateM)
+
+import Clash.Explicit.Prelude
+import Clash.Cores.Xilinx.DcFifo.Explicit
+import Clash.Netlist.Util (orNothing)
+
+
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+import Hedgehog (MonadGen, (===), property, forAll, Property)
+import qualified Hedgehog as H
+import Test.Tasty.Hedgehog (testPropertyNamed)
+import Test.Tasty (testGroup, TestTree)
+
+tests :: TestTree
+tests = testGroup "FIFO tests"
+  [ testPropertyNamed
+    "FIFO doesn't lose any data with small stalls"
+    "prop_noloss"
+    prop_noloss
+  , testPropertyNamed "FIFO preserves order" "prop_fifoOrder" prop_fifoOrder
+  ]
+
+intersperseStalls ::
+  -- | Data
+  [a] ->
+  -- | Stalls
+  [b] ->
+  [Either b a]
+intersperseStalls = go False where
+  go True (d:ds) s = Right d : go False ds s
+  go False ds (s:ss) = Left s : go True ds ss
+  go _ [] s = Left <$> s
+  go _ ds [] = Right <$> ds
+
+genScenario ::
+  (KnownNat n, MonadGen m) =>
+  Int -> Int ->
+  m ([BitVector n], [Either Int (BitVector n)], [Maybe Int])
+genScenario readMax writeMax = do
+  xs <- replicateM 10 genData
+  stallRead <- Gen.maybe (replicateM 10 $ Gen.int (Range.linear 0 readMax))
+  stallWrite <- Gen.maybe (replicateM 10 $ Gen.int (Range.linear 0 writeMax))
+  let iWrites = fromMaybe [] stallWrite
+  let readStalls = case stallRead of
+        Nothing -> []
+        Just i -> fmap Just i
+  pure (xs, intersperseStalls xs iWrites, L.intersperse Nothing readStalls)
+
+-- | Assert that everything in @xs@ is in @ys@ and has the same relative rank.
+orderFifo ::
+  Eq a =>
+  -- | @xs@
+  [a] ->
+  -- | @ys@
+  [a] ->
+  Bool
+orderFifo [] [] = True
+orderFifo (x:xs) (y:ys) | x == y = orderFifo xs ys
+orderFifo xs (_:ys) = orderFifo xs ys
+orderFifo _ _ = False
+
+-- | Generated stalls are large enough that we expect loss, but order should be
+-- preserved.
+prop_fifoOrder :: Property
+prop_fifoOrder = property $ do
+  (xs, wrIn, rdStalls) <- forAll $ genScenario 30 30
+  H.assert (orderFifo (throughFifo wrIn rdStalls) xs)
+
+-- | Generated stalls are small enough that we don't expect any data loss
+prop_noloss :: Property
+prop_noloss = property $ do
+  (xs, wrIn, rdStalls) <- forAll $ genScenario 12 12
+  throughFifo wrIn rdStalls === xs
+
+genData :: (KnownNat n, MonadGen m) => m (BitVector n)
+genData = Gen.enumBounded
+
+takeState ::
```

Comment:
What I was talking about is making a synthesizable version of this (and its counterpart), so we could build a synthesizable test. Randomization in HDL is hard, so I'd probably have the type signature read something like:

```haskell
type ReadLastCycle = Bool
type Stall = Bool
type ExpectedToRead n = BitVector n
type UnexpectedRead = Bool

fifoSampler ::
  Clock dom -> Reset dom -> Enable dom ->
  -- | Stall circuit? [note: probably coming from an LSFR, see `Clash.Examples`]
  Signal dom Stall ->
  -- | Signals from FIFO
  Signal dom (ResetBusy, Empty, DataCount depth, BitVector n) ->
  -- | Maybe output read from FIFO
  Signal dom (Maybe (BitVector n))
fifoSampler clk rst ena stalls inps =
  mealyB clk rst ena go False (stalls, inps)
 where
  go :: 
    ReadLastCycle ->
    (Stall, (ResetBusy, Empty, DataCount depth, BitVector n)) ->
    (ReadLastCycle, Maybe BitVector n)
  go = _
```

On the feed side you could generate a sequence to write to the FIFO (0, 1, 2..) so the type signature could look something like:

```haskell
type ReadLastCycle = Bool
type Stall = Bool
type ExpectedToRead n = BitVector n
type UnexpectedRead = Bool

fifoDriver ::
  Clock dom -> Reset dom -> Enable dom ->
  -- | Stall circuit? [note: probably coming from an LSFR, see `Clash.Examples`]
  Signal dom Stall ->
  -- | Signals from FIFO
  Signal dom (ResetBusy, Full, DataCount depth) ->
  -- | Maybe write input to FIFO
  Signal dom (Maybe (BitVector n))
fifoDriver clk rst ena stalls inps =
  mealyB clk rst ena go 0 (stalls, inps)
 where
  go :: 
    BitVector n ->
    (Stall, (ResetBusy, Full, DataCount depth)) ->
    (BitVector n, Maybe BitVector n)
  go = _
```

We can pair program a bit to get this working if you'd like.

Reply:
Ah! Okay, I will take a look + fix.

Reply:
Ok, this is done.